src/InputOutput/writers.jl

```diff
@@ -511,6 +531,118 @@ function write!(
     )
 end
 
+"""
+    write!(
+        writer::HDF5Writer,
+        values::DataLayouts.AbstractData,
+        name::AbstractString,
+        topology::Topologies.AbstractTopology,
+    )
+
+Write an object of type 'AbstractData' and name 'name' to the HDF5 file.
+
+The `values` should belong to a `Field` whose `space`'s topology is
+`topology(axes(field))`.
+"""
+function write!(
+    writer::HDF5Writer,
+    group,
+    values::DataLayouts.AbstractData,
+    name::AbstractString,
+    topology::Topologies.AbstractTopology,
+)
+    if topology isa Topologies.Topology2D &&
+       !(writer.context isa ClimaComms.SingletonCommsContext)
+        nelems = Topologies.nelems(topology)
+        (; local_elem_gidx) = topology
+        _write_mpi!(group, values, name; nelems, local_elem_gidx)
+    else
+        _write!(group, values, name)
+    end
+end
+
+function write_plain_array!(group, array::AbstractArray, name::AbstractString)
+    nd = ndims(array)
+    dims = size(array)
+    localidx = ntuple(d -> (:), nd)
+    dataset =
+        create_dataset(group, name, datatype(eltype(array)), dataspace(dims))
+    dataset[localidx...] = array
+    return dataset
+end
+
+"""
+    _write_mpi!(
+        writer::HDF5Writer,
+        data::DataLayouts.AbstractData,
+        name::AbstractString,
+        nelems,
+        local_elem_gidx
+    )
+
+This is an internal method, meant to be used for writing data layouts to the
+HDF5 file.
+
+This method should be used when
+
+```julia
+!(topology isa Topologies.Topology2D && !(writer.context isa ClimaComms.SingletonCommsContext))
+```
+"""
+function _write_mpi!(
```

Comment:
I haven't gone through the PR in details yet, but this caught my eye. When InputOutput writes other objects (including Fields and their associated data) it can directly handle the distributed case. Why is there a need to distinguish the MPI and non-MPI case here?

Also,
```
!(topology isa Topologies.Topology2D && !(writer.context isa ClimaComms.SingletonCommsContext))
```

The double negation makes this hard to understand. It seems that this says that the function should be used when the topology is not 2D and it is singleton, the opposite of what the function name suggests. Is this a typo?

Reply:
Okay, I looked at the existing code and think I have answered my own question about the need for two functions: they are not really handled directly and there are still if-else statements

Reply:
Yeah, I wish there was a better way, but you really need to know topology information to handle datalayout IO correctly. I split the method into two pieces in case either of them could be reused. For now, I'm only using these methods for the mask but we could eventually reuse them in other places to reduce some of the duplicate logic.

I think  happy to remove the double negative from the docs, I tend to avoid them, but I think I just did that to somewhat match the source code.

Reply:
I simplified the doc string.